core_new/src/message/common.rs

```diff
@@ -0,0 +1,163 @@
+    destination: ProgramId,
+    /// Message payload.
+    payload: Payload,
+    /// Message optional gas limit.
+    gas_limit: Option<GasLimit>,
+    /// Message value.
+    value: Value,
+    /// Message id replied on with exit code.
+    reply: Option<(MessageId, ExitCode)>,
+}
+
+impl Message {
+    /// Create new message.
+    pub fn new(
+        id: MessageId,
+        source: ProgramId,
+        destination: ProgramId,
+        payload: Payload,
+        gas_limit: Option<GasLimit>,
+        value: Value,
+        reply: Option<(MessageId, ExitCode)>,
+    ) -> Self {
+        Self {
+            id,
+            source,
+            destination,
+            payload,
+            gas_limit,
+            value,
+            reply,
+        }
+    }
+
+    /// Convert Message into gasless StoredMessage.
+    pub fn into_stored(self) -> StoredMessage {
+        StoredMessage::new(
+            self.id,
+            self.source,
+            self.destination,
+            self.payload,
+            self.value,
+            self.reply,
+        )
+    }
+
+    /// Message id.
+    pub fn id(&self) -> MessageId {
+        self.id
+    }
+
+    /// Message source.
+    pub fn source(&self) -> ProgramId {
+        self.source
+    }
+
+    /// Message destination.
+    pub fn destination(&self) -> ProgramId {
+        self.destination
+    }
+
+    /// Message payload reference.
+    pub fn payload(&self) -> &[u8] {
+        self.payload.as_ref()
+    }
+
+    /// Message optional gas limit.
+    pub fn gas_limit(&self) -> Option<GasLimit> {
+        self.gas_limit
+    }
+
+    /// Message value.
+    pub fn value(&self) -> Value {
+        self.value
+    }
+
+    /// Check if this message is reply.
+    pub fn is_reply(&self) -> bool {
+        self.reply.is_some()
+    }
+
+    /// Message id what this message replies to, if reply.
+    pub fn reply_to(&self) -> Option<MessageId> {
+        self.reply.map(|(id, _)| id)
+    }
+
+    /// Exit code of the message, if reply.
+    pub fn exit_code(&self) -> Option<ExitCode> {
+        self.reply.map(|(_, exit_code)| exit_code)
+    }
+}
+
+/// Message with entry point.
+#[derive(Clone, Debug, Eq, Hash, Ord, PartialEq, PartialOrd, Decode, Encode, TypeInfo)]
+pub struct Dispatch {
+    /// Entry point for the message.
+    kind: DispatchKind,
+    /// Message.
+    message: Message,
+}
+
+impl Dispatch {
+    /// Create new Dispatch.
+    pub fn new(kind: DispatchKind, message: Message) -> Self {
+        Self { kind, message }
+    }
+
+    /// Convert Dispatch into gasless StoredDispatch with empty previous context.
+    pub fn into_stored(self) -> StoredDispatch {
+        StoredDispatch::new(self.kind, self.message.into_stored(), None)
+    }
+
+    /// Entry point for the message.
+    pub fn kind(&self) -> DispatchKind {
+        self.kind
+    }
+
+    /// Dispatch message reference.
+    pub fn message(&self) -> &Message {
+        &self.message
+    }
+}
+
+impl Deref for Dispatch {
```

Comment:
I'm not sure that `deref`ed dispatch will look not confusing. We can use `message()` method to access the inner message.

Reply:
Why not? Anything associated to dispatch means message. So it seems logically that dispatch.id() == message.id() (inside the dispatch)